### DISCUSSION TITLE
State Store in Streaming Systems

### DISCUSSION THREAD

**Discussion by skyzh** [skyzh/skyzh-site#27] [General]
https://skyzh.dev/blog/2023-12-28-store-of-streaming-states/

**Comment by frankmcsherry** [comment]
This is generally a great post, and the information about DD and MZ is very sharp. There's one issue that I wanted to call out though!

You say, in the section "Overheads of Remote Accessing State", that

> In this case, remote access to Arrangements is inevitable in Differential Dataflow. 

However, there is *zero* remote access to arrangements in differential dataflow. The code literally does not exist. I think the misunderstanding can be traced back to 

> [...] the Arrangements of two tables A and B may be generated on different nodes (Node 1, 2), [...]

which supposes that arrangements are made on nodes. In fact, arrangements are *sharded across* nodes, by key. So the arrangements of tables A and B will have a fraction of their keyspace on Node 1, and another fraction on Node 2. As long as the sharding by key is consistent, each of Nodes 1 and 2 will have the right subsets of A and B locally, and can implement a join with no further data exchange or remote access.
